contracts/src/ERC20.sol

```diff
@@ -0,0 +1,351 @@
+     * @dev Sets the values for {name}, {symbol}, and {decimals}.
+     */
+    constructor(string memory name_, string memory symbol_, uint8 decimals_) {
+        OWNER = msg.sender;
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                DOMAIN_TYPE_SIGNATURE_HASH, keccak256(bytes(name_)), keccak256(bytes("1")), block.chainid, address(this)
+            )
+        );
+
+        name = name_;
+        symbol = symbol_;
+        decimals = decimals_;
+    }
+
+    modifier onlyOwner() {
+        if (msg.sender != OWNER) {
+            revert Unauthorized();
+        }
+        _;
+    }
+
+    /**
+     * @dev Creates `amount` tokens and assigns them to `account`, increasing
+     * the total supply. Can only be called by the owner.
+     *
+     * Emits a {Transfer} event with `from` set to the zero address.
+     *
+     * Requirements:
+     *
+     * - `account` cannot be the zero address.
+     */
+    function mint(address account, uint256 amount) external virtual onlyOwner {
+        _mint(account, amount);
+    }
+
+    /**
+     * @dev Destroys `amount` tokens from the account.
+     */
+    function burn(address account, uint256 amount) external virtual onlyOwner {
+        _burn(account, amount);
+    }
+
+    /**
+     * @dev See {IERC20-transfer}.
+     *
+     * Requirements:
+     *
+     * - `recipient` cannot be the zero address.
+     * - the caller must have a balance of at least `amount`.
+     */
+    function transfer(address recipient, uint256 amount) external virtual override returns (bool) {
+        _transfer(msg.sender, recipient, amount);
+        return true;
+    }
+
+    /**
+     * @dev See {IERC20-approve}.
+     *
+     * NOTE: Prefer the {increaseAllowance} and {decreaseAllowance} methods, as
+     * they aren't vulnerable to the frontrunning attack described here:
+     * https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+     * See {IERC20-approve}.
+     *
+     * NOTE: If `amount` is the maximum `uint256`, the allowance is not updated on
+     * `transferFrom`. This is semantically equivalent to an infinite approval.
+     *
+     * Requirements:
+     *
+     * - `spender` cannot be the zero address.
+     */
+    function approve(address spender, uint256 amount) external virtual override returns (bool) {
+        _approve(msg.sender, spender, amount);
+        return true;
+    }
+
+    /**
+     * @dev See {IERC20-transferFrom}.
+     *
+     * Emits an {Approval} event indicating the updated allowance. This is not
+     * required by the EIP. See the note at the beginning of {ERC20}.
+     *
+     * Requirements:
+     *
+     * - `sender` and `recipient` cannot be the zero address.
+     * - `sender` must have a balance of at least `amount`.
+     * - the caller must have allowance for ``sender``'s tokens of at least
+     * `amount`.
+     */
+    function transferFrom(address sender, address recipient, uint256 amount) external virtual override returns (bool) {
```

Comment:
I think we should move most of these external functions to a common library, to save on deployment costs. Especially since this code is going to be deployed hundreds of times. It will save on bridging fees.

Reply:
maybe we should just be deploying simple proxies for each token rather than a full erc20 contract each time? but it's not a big deal i guess, even hundreds of times of redeploys won't cost too much

Reply:
Yeah, maybe we can delegate these advanced features(allowance/permit) to something like https://github.com/Uniswap/permit2

Currently, the ERC20.sol is only 4KB so may not be a big issue.

Reply:
> maybe we should just be deploying simple proxies for each token rather than a full erc20 contract each time? but it's not a big deal i guess, even hundreds of times of redeploys won't cost too much

Yeah, I was thinking about this yesterday actually. It seems [ERC-6909](https://github.com/jtriley-eth/ERC-6909) is the emerging standard for multi-token contracts (replacing ERC-1155). Its being used in Uniswap-v4 And so I was thinking Gateway can maintain a single ERC-6909 contract, and then instantiate thin ERC20 proxies over it.

I am ambivalent about it.

Positives:
* Significant gas savings
* Simplified protocol as it would no longer involve agents (the contract would be owned by the gateway rather than the agent of the origin parachain).
* The ability to upgrade the multi-token contract, to add new features and fix bugs. This is not really possible in the existing design where we instantiate separate ERC20 contracts.

Negatives:
* Gav really wanted Agents to own the wrapped tokens, as its fits with his particular vision of XCM as a cross-chain protocol. So that may be a blocker.
* ERC-6909 is fairly [opinionated](https://ethereum-magicians.org/t/erc-6909-multi-token-standard/13891/36). For example, it doesn't support [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612) signed approvals, and instead has an approved operator mechanism (similar to ERC-777)

Reply:
can we make a decision here? maybe best to keep it simple and stick with existing standards?

Reply:
Yes, I am favor the current approach - just brought the idea up in case anyone had strong opinions.

@yrong lets go with the current approach, but with most of `ERC20.sol` factored out into a library `ERC20Lib.sol`

Reply:
Vincent, so which functions you'd like to move to a library?

I would assume the contract size may not be a big issue here and will not make a difference for the bridging fees?

 If we do want to slim the contract somehow, I'd prefer to remove all token permit/approve functions to make ERC20.sol minimal. 
 
 For all [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612), [EIP-1271](https://eips.ethereum.org/EIPS/eip-1271) features we delegate to a separate contract like [permit2](https://github.com/Uniswap/permit2). 
 
so WDYT?

Reply:
As per this permit2 [explainer](https://github.com/dragonfly-xyz/useful-solidity-patterns/tree/main/patterns/permit2), permit2 is really meant to be compatibility layer for older tokens that don't support EIP-2612. Its not meant to be a replacement for it.

And the user still has to manually approve the permit2 contract to spend the tokens. So there is an initial setup step that detracts from the UX experience.

Reply:
I don't think it will take long to introduce `ERC20Lib.sol`. The hardest part will be moving the token storage into a struct like https://github.com/aragon/zeppelin-solidity/blob/23540364330f6b109980933c032238dd4d1308d1/contracts/token/ERC20Lib.sol#L8

After that's done, I would suggest moving `_update` into the library first, as its the largest function